scripts/showcase/map_fastqs.py

```diff
@@ -3,22 +3,435 @@
+
+    return tracking_sheet_df
+
+
+def update_subject_objects(subject, output_path):
+    """
+    Given a list of subject objects, run the necessary internal functions to update attributes
+    Parameters
+    ----------
+    subject
+    output_path: Path
+
+    Returns
+    -------
+
+    """
+
+    for subject in subject:
+        # Set / create the output path for each subject
+        subject.set_output_path(output_path)
+
+
+def merge_sample_sheet_and_tracking_sheet(sample_sheet_df, metadata_df):
+    """
+    Merge sample sheet and tracking sheet
+    Parameters
+    ----------
+    sample_sheet_df: pd.DataFrame
+    metadata_df: pd.DataFrame
+
+    Returns
+    -------
+    merged_df: pd.DataFrame
+        =========    =======================================
+        RGID         i7Index.i5Index.Lane
+        RGSM         Sample Name as on the sample sheet
+        RGLB         Library ID (Set to Unknown Library)
+        Lane         Lane ID
+        Read1File    Path to Read 1 File
+        Read2File    Path to Read 2 File
+        # Extended columns
+        SampleID     Identifier of the sample, may be useful if we're running with top ups
+        Phenotype    Either 'tumor' or 'normal'
+
+    """
+    # Columns to keep
+
+    slimmed_metadata_df = metadata_df.filter(items=METADATA_COLUMNS)
+
+    merged_df = pd.merge(sample_sheet_df, slimmed_metadata_df,
+                         left_on="RGSM", right_on="Sample_ID (SampleSheet)",
+                         how="left")
+
+    # Ensure Phenotype is either 'tumor' or 'normal'
+    merged_df["Phenotype"] = merged_df["Phenotype"].astype(PHENOTYPES_DTYPE)
+
+    # Check for missing SampleIDs - Positive controls sometimes don't have a sample ID
+    if merged_df['SampleID'].isna().any():
+        logger.warning("Could not retrieve the SubjectID information for samples {}".format(
+            ', '.join(merged_df.query("SubjectID.isna()")['Sample_ID (SampleSheet)'].tolist())
+        ))
+
+    # Check for missing SubjectIDs
+    if merged_df['SubjectID'].isna().any():
+        logger.warning("Could not retrieve the SubjectID information for samples {}".format(
+            ', '.join(merged_df.query("SubjectID.isna()")['Sample_ID (SampleSheet)'].tolist())
+        ))
+
+    # Check for missing phenotypes in samples - could be from an invalid merge or bad name
+    if merged_df['Phenotype'].isna().any():
+        logger.warning("Could not retrieve the phenotype information for samples {}".format(
+            ', '.join(merged_df.query("Phenotype.isna()")['Sample_ID (SampleSheet)'].tolist())
+        ))
+
+    # Remove controls (samples that start with PTC or NTC)
+    # See the regex magic here: https://regex101.com/r/rROljA/1
+    # Str contains explanation here: https://stackoverflow.com/a/44335734/6946787
+    # '~' does the negating.
+    merged_df.query("~SampleID.str.contains(@CONTROL_REGEX_MATCH, regex=True)", inplace=True)
+
+    # Remove top ups
```

Comment:
Why remove top-ups?

Reply:
In theory top ups wouldn't be on the same run as their corresponding normal sample and so would need to be processed manually, rather than run a 5X tumour vs 30X normal WGS run?